src/UglyToad.PdfPig/DocumentLayoutAnalysis/RecursiveXYCut.cs

```diff
@@ -0,0 +1,411 @@
+            {
+                if ((words[i].BoundingBox.Bottom >= currentProj[0] && words[i].BoundingBox.Bottom <= currentProj[1])
+                    || (words[i].BoundingBox.Top >= currentProj[0] && words[i].BoundingBox.Top <= currentProj[1]))
+                {
+                    // it is overlapping 
+                    if (words[i].BoundingBox.Bottom >= currentProj[0]
+                        && words[i].BoundingBox.Bottom <= currentProj[1]
+                        && words[i].BoundingBox.Top > currentProj[1])
+                    {
+                        currentProj[1] = words[i].BoundingBox.Top;
+                    }
+                }
+                else
+                {
+                    // no overlap
+                    if (words[i].BoundingBox.Bottom - currentProj[1] <= domFontHeight)
+                    {
+                        // if gap too small -> don't cut
+                        // |____| |____|
+                        currentProj[1] = words[i].BoundingBox.Top;
+                    }
+                    else
+                    {
+                        // if gap big enough -> cut!
+                        // |____|   |   |____|
+                        if (i != wordsCount - 1) // will always add the last one after
+                        {
+                            projectionProfile.Add(currentProj);
+                            currentProj = new decimal[2] { words[i].BoundingBox.Bottom, words[i].BoundingBox.Top };
+                        }
+                    }
+                }
+                if (i == wordsCount - 1) projectionProfile.Add(currentProj);
+            }
+
+            if (projectionProfile.Count == 1)
+            {
+                if (level >= 1)
+                {
+                    return leef;
+                }
+                else
+                {
+                    level++;
+                }
+            }
+
+            var newLeefsEnums = projectionProfile.Select(p =>
+                leef.Words.Where(w => w.BoundingBox.Bottom >= p[0] && w.BoundingBox.Top <= p[1]));
+            var newLeefs = newLeefsEnums.Where(e => e.Count() > 0).Select(e => new XYLeef(e));
+            var newNodes = newLeefs.Select(l => VerticalCut(l, minimumWidth,
+                dominantFontWidthFunc, dominantFontHeightFunc, level)).ToList();
+
+            var lost = leef.Words.Except(newLeefsEnums.SelectMany(x => x)).Where(x => !string.IsNullOrWhiteSpace(x.Text)).ToList();
+            if (lost.Count > 0)
+            {
+                newNodes.AddRange(lost.Select(w => new XYLeef(w)));
+            }
+            return new XYNode(newNodes);
+        }
+    }
+
+    /// <summary>
+    /// A Node used in the <see cref="RecursiveXYCut"/> algorithm.
+    /// </summary>
+    public class XYNode
+    {
+        /// <summary>
+        /// Returns true if this node is a leef, false otherwise.
+        /// </summary>
+        public virtual bool IsLeef => false;
```

Comment:
I think the naming should be `Leaf` here and in other places unless leef has a special meaning in this context?

Reply:
Hehe makes sense, it's a typo indeed